In-Place Pod Vertical Scaling feature

vinaykul opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
/kind api-change
> /kind bug
> /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**: 
This PR implements the API change for [In-Place Pod Vertical Scaling KEP](https://github.com/kubernetes/enhancements/blob/8bb838caac1251be03785c8974ad72a3e73af629/keps/sig-node/20181106-in-place-update-of-pod-resources.md) , the [Kubelet CRI KEP](https://github.com/kubernetes/enhancements/blob/8bb838caac1251be03785c8974ad72a3e73af629/keps/sig-node/20191025-kubelet-container-resources-cri-api-changes.md) to support in-place pod vertical scaling feature , and the core implementation of in-place pod vertical scaling feature.

It adds new fields named ResourcesAllocated and ResizePolicy to  Pod's Container type, and Resources field to ContainerStatus type. It also adds a new admission controller that limits the ability to change ResourcesAllocated field to system-node account. This API change enables users to scale a Pod's CPU and memory resources up or down.

The CRI change enables runtimes to report currently applied resource limits for a container, and adds support for Windows containers.

The core implementation commit implements the [In-Place Pod Vertical Scaling KEP](https://github.com/kubernetes/enhancements/blob/8bb838caac1251be03785c8974ad72a3e73af629/keps/sig-node/20181106-in-place-update-of-pod-resources.md)

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

**Special notes for your reviewer**: 
1. The commit named _'InPlace Pod Vertical Scaling feature - API change'_ implements the API change for In-Place Pod Vertical Scaling.
2. The commit named _'InPlace Pod Vertical Scaling feature - Kubelet CRI change'_ implements the CRI changes to allow runtimes to report resource limits.
3. The commit named _'InPlace Pod Vertical Scaling feature - core implementation'_ is the basic implementation of In-Place Pod Vertical Scaling feature that performs the core resizing and resize policy handling.

Switching scheduler, kubelet, and apiserver code (limit range, resource quota) is coming in a future commit.

**Does this PR introduce a user-facing change?**: Yes
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
1. A new field named ResourcesAllocated (type v1.ResourceList) is added to Container struct.
2. A new field named ResizePolicy is added to Container struct.
3. A new field named Resources (type v1.ResourceRequirements) is added to ContainerStatus struct.
4. A new admission controller named 'PodResourceAllocation' is introduced to limit access to modification of ResourcesAllocated field.

e.g: 
root@skibum:~/vInPlacePodVerticalScaling# cat ~/YML/1pod.yaml 
apiVersion: v1
kind: Pod
metadata:
  name: 1pod
spec:
  containers:
  - name: stress
    image: skiibum/ubuntu-stress:18.10
    resources:
      limits:
        cpu: "1"
        memory: "1Gi"
      requests:
        cpu: "1"
        memory: "1Gi"
    resizePolicy:
    - resourceName: cpu
      policy: NoRestart
    - resourceName: memory
      policy: RestartContainer


```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: https://github.com/vinaykul/enhancements/blob/cad896de8ac1ab5f6ed72bc4ab06c76ce4438543/keps/sig-node/20181106-in-place-update-of-pod-resources.md
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
In-Place Update of Pod Resources KEP:
https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/20181106-in-place-update-of-pod-resources.md

Container Resources CRI API Changes for Pod Vertical Scaling KEP:
https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/20191025-kubelet-container-resources-cri-api-changes.md
```

thockin commented on pkg/apis/core/types.go:
```diff
@@ -1966,6 +1966,28 @@ const (
+	RestartContainer ContainerResizePolicy = "RestartContainer"
+)
+
+// ResizePolicy represents the resource resize policy for a single container.
```
Please be more precise about the meaning of this - is it a requirement that this container not be restarted or is it saying that a restart is not required (but may still happen if the kubelet needs to)?  Or something else.  I think this distinction may influence the constant strings above.

E.g. I *assume* this is a statement of capability.  "I can be resized in-place without restart".  That would lead me to look for something like:

```
resizePolicy:
  - resource: "cpu"
    restart: "NotRequired"
  - resource: "memory"
    restart: Required
```

What *other* policies do you think might one day be needed here?

vinaykul replied:
NoRestart tells Kubelet to call UpdateContainerResources CRI API to resize the resource - doing so does not restart the container. RestartContaienr policy on the other hand tell Kubelet to stop and start the container with new resources, and the use-case for this was some legacy java apps using -xmxN flag which are unable to use the increased memory without restarting.

We had RestartPod as a third option at one point, but removed it because we couldn't find a real-world use case for it other than the notion that someone might want to restart the entire pod in-place (to rerun the init containers).

I'll add more details to the comments.

vinaykul replied:
Are you thinking this should just be a binary/bool value considering we have just two values for the foreseeable future?

thockin replied:
Hi, I found this half-completed, I don't know if it is still alive.

My point was to clarify whether "no restart" is contract or guidance.  If the kubelet, for some reason, needed to restart the pod, would that be a violation of NoRestart or just an unfortunate, but legal, edge case?

vinaykul replied:
I've updated [documentation](https://github.com/vinaykul/kubernetes/pull/1/commits/5126b9e1e93d2530ac820981c93986d64e46faa3#diff-4143fe2cdf42f1748648a9b80aa2ba07R1975) in the code to clarify that container resize policy is a guidance by stating that kubelet should resize the container without restarting it _if possible_ . I hope this is sufficient language.

thockin commented on pkg/apis/core/v1/defaults.go:
```diff
@@ -176,6 +176,36 @@ func SetDefaults_Pod(obj *v1.Pod) {
+		if utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling) {
+			// set resources allocated equal to requests, if not specified
+			if obj.Spec.Containers[i].Resources.Requests != nil {
+				if obj.Spec.Containers[i].ResourcesAllocated == nil {
```
What if the create operation specified ResourcesAllocated?  Then we don't initialize them?

vinaykul replied:
In admission plugin, we verify that it matches Resources.Requests, and if not reject the operation.

On the KEP, we had discussed the possibility of user requesting ResourcesAllocated different (typically lower, minimum required) from ideal desired resources in Requests, but agreed to take that up as a future enhancement.

vinaykul replied:
As scoped in this KEP, my implementation will fail validation if user specifies ResourcesAllocated that is different from Requests.

thockin commented on pkg/apis/core/v1/defaults.go:
```diff
@@ -176,6 +176,36 @@ func SetDefaults_Pod(obj *v1.Pod) {
+			for _, p := range obj.Spec.Containers[i].ResizePolicy {
+				resources[p.ResourceName] = true
+			}
+			if _, found := resources[v1.ResourceCPU]; !found {
```
Comment why this is coded to support CPU and memory and nothing else?  What happens if some OTHER resource is provided (e.g. GPUs).  It doesn't get listed in resizePolicy?

vinaykul replied:
We currently plan to support in-place resize only for CPU and memory resources. So I restrict setting the policy only for cpu, memory in the admission plugin. But good point, maybe I should default all of them to NoRestart - we would have done that anyways if we did support other resource types. Let me think about this.

vinaykul replied:
The current implementation defaults both CPU and memory to NoRestart.

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -2609,6 +2610,42 @@ func validatePullPolicy(policy core.PullPolicy, fldPath *field.Path) field.Error
+func validateResizePolicy(policyList []core.ResizePolicy, fldPath *field.Path) field.ErrorList {
+	allErrors := field.ErrorList{}
+
+	if !utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling) && len(policyList) > 0 {
```
generally we do not want to check gates in validation - that happens before we ever get here.  See DropDisabledPodFields()

vinaykul replied:
thanks! i'll fix this.

vinaykul replied:
fixed in commit https://github.com/vinaykul/kubernetes/pull/1/commits/5126b9e1e93d2530ac820981c93986d64e46faa3

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -2609,6 +2610,42 @@ func validatePullPolicy(policy core.PullPolicy, fldPath *field.Path) field.Error
+	resources := make(map[core.ResourceName]bool)
+	for _, p := range policyList {
+		if _, found := resources[p.ResourceName]; found {
+			allErrors = append(allErrors, field.Invalid(fldPath, p.ResourceName, "resource must not be repeated"))
```
field.Duplicate()

vinaykul replied:
fixed in commit https://github.com/vinaykul/kubernetes/pull/1/commits/5126b9e1e93d2530ac820981c93986d64e46faa3

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -3735,12 +3773,26 @@ func ValidatePodUpdate(newPod, oldPod *core.Pod) field.ErrorList {
 	}
 
+	if utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling) {
+		// reject QOS change attempt
```
Unrelated?

vinaykul replied:
legit. we will allow user to change resources but can't make a burstable class pod guaranteed and so on.

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -3735,12 +3773,26 @@ func ValidatePodUpdate(newPod, oldPod *core.Pod) field.ErrorList {
+		if utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling) {
+			// allow Resources and ResourcesAllocated fields to be mutable
+			container.Resources = oldPod.Spec.Containers[ix].Resources
+			container.ResourcesAllocated = oldPod.Spec.Containers[ix].ResourcesAllocated
```
Does this mean a user can change the resourcesAllocated?

vinaykul replied:
user cannot - admission plugin will yell. kubelet can.

dashpole commented:
/cc

dchen1107 commented:
/cc

dashpole commented:
lgtm, this matches what I expect from the KEP.

vinaykul commented:
@thockin @liggitt

Apologies for the delay, I was pulled off to work on another priority item due to a couple of our delayed coming back from China after they went there for the new year holidays. 

Please review update https://github.com/vinaykul/kubernetes/pull/1/commits/a419856f5d75741e3735954bd260a76792c5dcce , it should address your comments.

thockin commented on pkg/apis/core/types.go:
```diff
@@ -2053,6 +2080,12 @@ type Container struct {
 	Resources ResourceRequirements
+	// Node compute resources allocated to the container.
+	// +optional
+	ResourcesAllocated ResourceList
```
I am still confused why this is spec?

vinaykul replied:
This is very much alive - I'm a couple of weeks away from having a mostly complete implementation out for review, and also updated API code to move some of the validation earlier into admission stage.

NoRestart/RestartContainer above is a guidance, and lets us decide whether to stop/start container or just call UpdateContainerResources API. The runtime may still choose to restart - we don't have control over it, and no explicit documentation telling runtime not to restart.

ResourcesAllocated was added to Spec to track resources that node has agreed reserve for the pod, and make this info visible to the user. 

We had discussed the alternative of storing this information locally on the node, but I was told that Borg had encountered issues with such approach, and they want to [avoid checkpointing on the node](https://github.com/kubernetes/enhancements/pull/686#discussion_r305090556), and thought it best to have Spec as the source of truth. We had [also looked at relying on PodStatus](https://github.com/kubernetes/enhancements/pull/686#discussion_r281825929), but the Status, per API conventions, isn't allowed store any state that cant be reconstructed through observations.

Commit:
InPlace Pod Vertical Scaling feature - API change
1. Add ResourcesAllocated and ResizePolicy fields to Container struct.
2. Add Resources field to ContainerStatus struct.
3. Add a new admission controller named PodResourceAllocation for setting defaults and validation.
4. Address code-review items:
   a. Clarify meaning of container resize policy that it is a guidance from user, not a guarantee by k8s.
   b. Handle updates from clients unaware of the new fields in pod spec.
   c. Use DropDisabledPodFields to avoid feature-gate checks in validation where possible.
KEP: /enhancements/keps/sig-node/20181106-in-place-update-of-pod-resources.md

Commit:
InPlace Pod Vertical Scaling feature - generated code from API change

Commit:
InPlace Pod Vertical Scaling feature - Kubelet CRI change
1. Add ContainerResources message that supports both Linux and Windows resources.
2. Add ContainerResources to ContainerStatus CRI API.
3. Modify UpdateContainerResources CRI API to use ContainerResources.
4. Implement handling ContainerStatus response for runtimes that don't support resources query.
KEP: /enhancements/keps/sig-node/20191025-kubelet-container-resources-cri-api-changes.md

Commit:
InPlace Pod Vertical Scaling feature - core implementation
1. Handle patching of ResourcesAllocated fields of a resized pod in Kubelet's syncPod routine.
2. Handle container limits updation in runtime manager's SyncPod routine.
3. Add new pods after testing for admissibility using ResourcesAllocated field values.
KEP: /enhancements/keps/sig-node/20181106-in-place-update-of-pod-resources.md

dashpole commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -1564,6 +1571,51 @@ func (kl *Kubelet) convertToAPIContainerStatuses(pod *v1.Pod, podStatus *kubecon
+						limits[v1.ResourceEphemeralStorage] = ephemeralStorage.DeepCopy()
+					}
+				}
+				// A running container's status.requests always comes from spec.resourcesAllocated
```
Shouldn't requests also come from the container runtime?  Requests are used for CPU shares IIRC.

vinaykul replied:
They should but I hit a corner-case I was unable to resolve. For millicpu requests of 1 vs 2, they both get a floor cpu.shares of 2. So when runtime reports back 2, I cannot tell whether it is 1m or 2m. Is it OK to report 2m even though user requested 1m? Maybe it is.. the container is going to get minimum shares of 2 regardless of 1m or 2m ask.

Even if I didn't do the above, there is another bug here - I should be updating ContainerStatus.Resources.Requests only after successful updateContainerLimits for all containers. I'll fix that, and revisit if I can use cpu.shares reported by runtime to reverse-compute cpu requests.

vinaykul replied:
Switched to getting Requests from cpu.shares in commit [2d88cb1](https://github.com/vinaykul/kubernetes/pull/1/commits/2d88cb142cb73073aa187edb2c8b6535ca416b8c). Seems to work fine.

dashpole commented on pkg/kubelet/cm/cgroup_manager_linux.go:
```diff
@@ -777,3 +778,82 @@ func (m *cgroupManagerImpl) GetResourceStats(name CgroupName) (*ResourceStats, e
+}
+
+// Set the cpu quota, cpu period, and cpu shares applied to the cgroup
+func (m *cgroupManagerImpl) SetCgroupCpuLimit(name CgroupName, cpuQuota *int64, cpuPeriod, cpuShares *uint64) error {
```
nit: this is more than the limit, as it is adjusting shares as well.

vinaykul replied:
Fixed. Renamed functions for greater clarity.

dashpole commented on pkg/kubelet/kubelet.go:
```diff
@@ -1635,6 +1646,59 @@ func (kl *Kubelet) syncPod(o syncPodOptions) error {
+	}
+	containersPatchData = strings.TrimRight(containersPatchData, ",")
+	patchData := fmt.Sprintf(`{"spec":{"containers":[%s]}}`, containersPatchData)
+	return true, patchData
```
We have a race condition here, but i don't remember how we agreed to solve it.  For example, if we called canResizePod on pod A and B, such that the ordering was:

1. canResizePod(A) => true
2. canResizePod(B) => true
3. Patch(A)
4. Patch(B)

This could pass even if there was only room for 1 of the pods, since A is not included in kl.GetActivePods() during canResizePod(B).

At minimum, I think we need to update GetActivePods here.  Or am I misremembering our conclusion?

vinaykul replied:
I've tested this for add vs resize race after coding up the switch-over from Requests to ResourcesAllocated (next change - hopefully by eod Monday). I think the mutex should serialize race between two resizes as well.

So in the above case, the next GetActivePods (for pod B( should use updated ResourcesAllocated values of Pod A if the resize was accepted. But let me specifically test that case and verify..

vinaykul replied:
I tested the case of concurrent updates, and the mutex mechanism holds up well.

dashpole replied:
Where do we update the active pods?

dashpole replied:
The other thing I was trying to reason through at sig-node is whether the update needs to be written back to the API Server while holding the lock.  Take the following example:

1. canResizePod(A) decrease by 100m => true
2. updateActivePods(A) decrease by 100m
3. canResizePod(B) increase by 100m => true, since A's requests are lowered
4. updateActivePods(B) increase by 100m
5. Patch(B)
6. Update (B) in the CRI
6. ... Patch(A) hangs for a while ... Update(A) in the CRI doesn't happen.

In this case, B would have a higher allocated even though A has not been patched to be lower.  An observer that did `kubectl describe no` would see resource allocated > allocatable.  We could decide that that is OK, since we _really_ just care about changes sent to the CRI, and the Patch() doesn't change that.

Does that make sense?

vinaykul replied:
> Where do we update the active pods?

In canResizePod function after successful admit of resized pod when we assign ```container.ResourcesAllocated[rName] = rQuantity```

Here's what it looks like when we slow it down (below code) and play it (log down below) . 16 cpu allocatable, pod1 resizes first from 2 cpu to 10 cpu, pod2 resizes from 3 cpu to 6 cpu, but does not get it - active pods has been updated when pod2 pulls up active pods even though pod1 apiserver patch happens much later.

```
diff --git a/pkg/kubelet/kubelet.go b/pkg/kubelet/kubelet.go
index 7c0c0cc..ecf9295 100644
--- a/pkg/kubelet/kubelet.go
+++ b/pkg/kubelet/kubelet.go
@@ -122,6 +122,7 @@ import (
        "k8s.io/kubernetes/pkg/volume/util/hostutil"
        "k8s.io/kubernetes/pkg/volume/util/subpath"
        "k8s.io/kubernetes/pkg/volume/util/volumepathhandler"
+"k8s.io/kubernetes/pkg/api/v1/resource"
 )
 
 const (
@@ -1648,10 +1649,12 @@ func (kl *Kubelet) syncPod(o syncPodOptions) error {
 
 func (kl *Kubelet) canResizePod(pod *v1.Pod) (bool, string) {
        var otherActivePods []*v1.Pod
+klog.Infof(">>__DBG_canResizePod_ENTER: pod=%s. Allocations: %+v\n", pod.Name, resource.PodResourceAllocations(pod))
        kl.podResizeMutex.Lock()
        defer kl.podResizeMutex.Unlock()
        activePods := kl.GetActivePods()
        for _, p := range activePods {
+klog.Infof(">>__DBG_canResizePod_ACTIVE_POD: pod=%s. active_pod=%s, Allocations: %+v\n", pod.Name, p.Name, resource.PodResourceAllocations(p))
                if p.UID != pod.UID {
                        otherActivePods = append(otherActivePods, p)
                }
@@ -1659,6 +1662,7 @@ func (kl *Kubelet) canResizePod(pod *v1.Pod) (bool, string) {
        if ok, failReason, failMessage := kl.canAdmitPod(otherActivePods, pod); !ok {
                // Log reason and return. Let the next sync iteration retry the resize
                klog.V(2).Infof("Pod '%s' resize cannot be accommodated. Reason: '%s' Message: '%s'", pod.Name, failReason, failMessage)
+klog.Infof(">>__DBG_canResizePod_RESIZE_ADMIT_FAILED: pod=%s. reason: %s, message: %s\n", pod.Name, failReason, failMessage)
                return false, ""
        }
 
@@ -1672,6 +1676,8 @@ func (kl *Kubelet) canResizePod(pod *v1.Pod) (bool, string) {
                resourcesPatchData = strings.TrimRight(resourcesPatchData, ",")
                containersPatchData += fmt.Sprintf(`{"name":"%s","resourcesAllocated":{%s}},`, container.Name, resourcesPatchData)
        }
+klog.Infof(">>__DBG_canResizePod_RESIZE_PATCH_CREATED: pod=%s. Allocations: %+v\n", pod.Name, resource.PodResourceAllocations(pod))
+time.Sleep(30 * time.Second) // CONCURRENCY_CREATOR
        containersPatchData = strings.TrimRight(containersPatchData, ",")
        patchData := fmt.Sprintf(`{"spec":{"containers":[%s]}}`, containersPatchData)
        return true, patchData
@@ -1690,10 +1696,13 @@ func (kl *Kubelet) handlePodResourcesResize(pod *v1.Pod) {
        }
 
        if fit, patchData := kl.canResizePod(pod); fit {
+klog.Infof(">>__DBG_handlePodResourcesResize_PATCHING_APISERVER: pod=%s\n", pod.Name)
+if pod.Name == "pod1" { time.Sleep(30 * time.Second) } // APISERVER_BUSY_EMULATOR
                _, patchError := kl.kubeClient.CoreV1().Pods(pod.Namespace).Patch(context.TODO(), pod.Name, types.StrategicMergePatchType, []byte(patchData), metav1.PatchOptions{})
                if patchError != nil {
                        klog.Errorf("Failed to patch ResourcesAllocated values for pod %s: %+v\n", pod.Name, patchError)
                }
+klog.Infof(">>__DBG_handlePodResourcesResize_PATCHING_APISERVER_DONE: pod=%s\n", pod.Name)
        }
        return
```
---------

```
root@fw0000358:~/jun01_vkInPlacePodVS_v1_19_dev/pkg/kubelet# cat /tmp/kubelet.log | grep __DBG_
I0609 21:44:02.198922   30286 kubelet.go:1652] >>__DBG_canResizePod_ENTER: pod=pod1. Allocations: map[cpu:{i:{value:2 scale:0} d:{Dec:<nil>} s:2 Format:DecimalSI}]
I0609 21:44:02.198966   30286 kubelet.go:1657] >>__DBG_canResizePod_ACTIVE_POD: pod=pod1. active_pod=kube-dns-747b6894bc-h8d95, Allocations: map[cpu:{i:{value:260 scale:-3} d:{Dec:<nil>} s: Format:DecimalSI} memory:{i:{value:115343360 scale:0} d:{Dec:<nil>} s: Format:BinarySI}]
I0609 21:44:02.198999   30286 kubelet.go:1657] >>__DBG_canResizePod_ACTIVE_POD: pod=pod1. active_pod=pod1, Allocations: map[cpu:{i:{value:2 scale:0} d:{Dec:<nil>} s:2 Format:DecimalSI}]
I0609 21:44:02.199025   30286 kubelet.go:1657] >>__DBG_canResizePod_ACTIVE_POD: pod=pod1. active_pod=pod2, Allocations: map[cpu:{i:{value:3 scale:0} d:{Dec:<nil>} s:3 Format:DecimalSI}]
I0609 21:44:02.200486   30286 kubelet.go:1679] >>__DBG_canResizePod_RESIZE_PATCH_CREATED: pod=pod1. Allocations: map[cpu:{i:{value:10 scale:0} d:{Dec:<nil>} s:10 Format:DecimalSI}]
I0609 21:44:09.774348   30286 kubelet.go:1652] >>__DBG_canResizePod_ENTER: pod=pod2. Allocations: map[cpu:{i:{value:3 scale:0} d:{Dec:<nil>} s:3 Format:DecimalSI}]
I0609 21:44:32.200830   30286 kubelet.go:1699] >>__DBG_handlePodResourcesResize_PATCHING_APISERVER: pod=pod1
I0609 21:44:32.200889   30286 kubelet.go:1657] >>__DBG_canResizePod_ACTIVE_POD: pod=pod2. active_pod=pod2, Allocations: map[cpu:{i:{value:3 scale:0} d:{Dec:<nil>} s:3 Format:DecimalSI}]
I0609 21:44:32.200945   30286 kubelet.go:1657] >>__DBG_canResizePod_ACTIVE_POD: pod=pod2. active_pod=kube-dns-747b6894bc-h8d95, Allocations: map[cpu:{i:{value:260 scale:-3} d:{Dec:<nil>} s: Format:DecimalSI} memory:{i:{value:115343360 scale:0} d:{Dec:<nil>} s: Format:BinarySI}]
I0609 21:44:32.200986   30286 kubelet.go:1657] >>__DBG_canResizePod_ACTIVE_POD: pod=pod2. active_pod=pod1, Allocations: map[cpu:{i:{value:10 scale:0} d:{Dec:<nil>} s:10 Format:DecimalSI}]
I0609 21:44:32.202919   30286 kubelet.go:1665] >>__DBG_canResizePod_RESIZE_ADMIT_FAILED: pod=pod2. reason: OutOfcpu, message: Node didn't have enough resource: cpu, requested: 6000, used: 10260, capacity: 16000
I0609 21:44:32.831605   30286 kubelet.go:1652] >>__DBG_canResizePod_ENTER: pod=pod2. Allocations: map[cpu:{i:{value:3 scale:0} d:{Dec:<nil>} s:3 Format:DecimalSI}]
I0609 21:44:32.831646   30286 kubelet.go:1657] >>__DBG_canResizePod_ACTIVE_POD: pod=pod2. active_pod=pod2, Allocations: map[cpu:{i:{value:3 scale:0} d:{Dec:<nil>} s:3 Format:DecimalSI}]
I0609 21:44:32.831674   30286 kubelet.go:1657] >>__DBG_canResizePod_ACTIVE_POD: pod=pod2. active_pod=kube-dns-747b6894bc-h8d95, Allocations: map[cpu:{i:{value:260 scale:-3} d:{Dec:<nil>} s: Format:DecimalSI} memory:{i:{value:115343360 scale:0} d:{Dec:<nil>} s: Format:BinarySI}]
I0609 21:44:32.831703   30286 kubelet.go:1657] >>__DBG_canResizePod_ACTIVE_POD: pod=pod2. active_pod=pod1, Allocations: map[cpu:{i:{value:10 scale:0} d:{Dec:<nil>} s:10 Format:DecimalSI}]
I0609 21:44:32.833177   30286 kubelet.go:1665] >>__DBG_canResizePod_RESIZE_ADMIT_FAILED: pod=pod2. reason: OutOfcpu, message: Node didn't have enough resource: cpu, requested: 6000, used: 10260, capacity: 16000
I0609 21:45:02.208825   30286 kubelet.go:1705] >>__DBG_handlePodResourcesResize_PATCHING_APISERVER_DONE: pod=pod1
I0609 21:45:36.642618   30286 kubelet.go:1652] >>__DBG_canResizePod_ENTER: pod=pod2. Allocations: map[cpu:{i:{value:3 scale:0} d:{Dec:<nil>} s:3 Format:DecimalSI}]
I0609 21:45:36.642658   30286 kubelet.go:1657] >>__DBG_canResizePod_ACTIVE_POD: pod=pod2. active_pod=kube-dns-747b6894bc-h8d95, Allocations: map[cpu:{i:{value:260 scale:-3} d:{Dec:<nil>} s: Format:DecimalSI} memory:{i:{value:115343360 scale:0} d:{Dec:<nil>} s: Format:BinarySI}]
I0609 21:45:36.642690   30286 kubelet.go:1657] >>__DBG_canResizePod_ACTIVE_POD: pod=pod2. active_pod=pod1, Allocations: map[cpu:{i:{value:10 scale:0} d:{Dec:<nil>} s:10 Format:DecimalSI}]
I0609 21:45:36.642712   30286 kubelet.go:1657] >>__DBG_canResizePod_ACTIVE_POD: pod=pod2. active_pod=pod2, Allocations: map[cpu:{i:{value:3 scale:0} d:{Dec:<nil>} s:3 Format:DecimalSI}]
I0609 21:45:36.644603   30286 kubelet.go:1665] >>__DBG_canResizePod_RESIZE_ADMIT_FAILED: pod=pod2. reason: OutOfcpu, message: Node didn't have enough resource: cpu, requested: 6000, used: 10260, capacity: 16000
root@fw0000358:~/jun01_vkInPlacePodVS_v1_19_dev/pkg/kubelet#
```

vinaykul replied:
About your scenario, yes in that case, kubectl user will see allocated > allocatable. I think that's ok - it would be a transient UX problem. 

But CRI update being held up may be an issue for the system... for that duration, we would be oversubscribed. At some point I did put a TODO marker in my code to possibly use a goroutine for apiserver patch op. Not sure why i dropped it.. i think i should add that back despite some added code complexity.

In either case, unless I missed something, we don't need to suffer the lock across PATCH call to apiserver. Concur?

vinaykul replied:
Regarding you scenario, I now recall the resiliency I built into it - pod B races with Pod A and calls CRI to increase cpu.shares before pod A has called CRI relinquishing its shares, and we are at system allocatable limits, I expected updateContainerResources call to fail. (We handle that by aborting and retrying later - I tested with failure injection).

However when I tried this scenario, I was surprised to find that updateContainerResources actually succeeded, and the sum of cgroup cpu.shares for Pod A + pod B exceeded the system max CPU capacity!  I was expecting this to fail - I'll dig into why cgroup allows this for shares.

But we should be OK even if a later pod races as long as cgroup gracefully handles the excess, even if it does so unfairly for a transient period. We don't really need the complexity of serializing the PATCH calls in a goroutine.

dashpole replied:
If we can end up with allocated > allocatable, that means a kubelet restart will result in one of the pods being rejected, as it reads pods from the source to admit them.

dashpole replied:
Also, WRT `container.ResourcesAllocated[rName] = rQuantity` updating active pods, I'd prefer explicitly calling podManager.UpdatePod() (or whatever the function name is) to make this explicit.  The podManager GetActivePods() function is meant to be read-only, but we don't do deep copies to enforce that.

vinaykul replied:
> If we can end up with allocated > allocatable, that means a kubelet restart will result in one of the pods being rejected, as it reads pods from the source to admit them.

good point. we have to patch holding the lock, ugh. i'll look into it after i'm done with resource quota and limit range, and will also fix the explicit update.

dashpole replied:
Sounds good.  We have tests that test the throughput of pod startup.  If it has a significant negative impact, we can consider the approach you currently have, and document the potential issues.

vinaykul replied:
Commit https://github.com/vinaykul/kubernetes/pull/1/commits/92ffa912382261445a48781ad59ee168390e0268 - Modified resource quota and limitranger plugins to handle updates to Requests and Limits, and modified kubelet to hold the lock across patch - it addresses the kubelet restart issue that you identified. Thanks!

dashpole commented on pkg/kubelet/kubelet.go:
```diff
@@ -1635,6 +1646,59 @@ func (kl *Kubelet) syncPod(o syncPodOptions) error {
+	}
+
+	if fit, patchData := kl.canResizePod(pod); fit {
+		// TODO: go routine for Patch
```
Are there any problems if we update the actual container requests and limits without updating the Pod object?  E.g. if the kubelet was heavily rate limited or the API Server is unavailable.

vinaykul replied:
From the node perspective, no. We update pod's resource allocation locally if it fits, and proceed to asking the runtime to update resources. If everything goes well, containers are at desired state - and we have successfully allocated  the pod what it requested.

But from pod's perspective, VPA may decide that we have not been able to resize successfully, and evict it depending on policy configuration.

dashpole commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -829,9 +920,147 @@ func (m *kubeGenericRuntimeManager) SyncPod(pod *v1.Pod, podStatus *kubecontaine
 }
 
+func (m *kubeGenericRuntimeManager) updateContainerLimits(pod *v1.Pod, podStatus *kubecontainer.PodStatus, resourceName v1.ResourceName, containersToUpdate []containerToUpdateInfo) error {
+	var updatedLimitsMap map[kubecontainer.ContainerID]struct {
```
nit: this var statement is unnecessary.

vinaykul replied:
Fixed. Cleaned up a bunch of code in this function.

dashpole commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -829,9 +920,147 @@ func (m *kubeGenericRuntimeManager) SyncPod(pod *v1.Pod, podStatus *kubecontaine
+		cStatus *kubecontainer.ContainerStatus
+		rLimits v1.ResourceList
+	}
+	updatedLimitsMap = make(map[kubecontainer.ContainerID]struct {
```
maybe define this struct in the function so we don't have to keep writing it out here and below

vinaykul replied:
Done in commit https://github.com/vinaykul/kubernetes/pull/1/commits/2d88cb142cb73073aa187edb2c8b6535ca416b8c

dashpole commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -829,9 +920,147 @@ func (m *kubeGenericRuntimeManager) SyncPod(pod *v1.Pod, podStatus *kubecontaine
+			return err
+		}
+		cInfo.apiContainerStatus.Resources.Limits = container.Resources.Limits
+		// NOTE: Ideally, cpu and memory resource limits come from ContainerStatus CRI API query.
```
just to confirm, this means that a failed update of container limits still appears as a success to observers of the pod object?

vinaykul replied:
No, if updateContainerResources returns failure at any point, we abort any subsequent updates and exit, and let the next SyncPod attempt retry this at a later time. It's only when all of the planned updates have succeeded that we execute the below GetPodStatus, update PLEG cache code. That update gets propagated to v1 ContainerStatus that user can see.

Commit:
InPlace Pod Vertical Scaling feature - review fixes to core implementation
1. Generate CPU Request API status information from runtime's cpu.shares report.
2. Rename pod cgrpoup config get/set functions for functional clarity, and simplify update container resources function.
KEP: /enhancements/keps/sig-node/20181106-in-place-update-of-pod-resources.md

Commit:
InPlace Pod Vertical Scaling feature - core implementation
1. Switch to using container.ResourcesAllocated instead of container.Resources.Requests for resource allocation/usage calculations.
2. Add support for ResourcesAllocated to kubectl describe with pod and node resources.
KEP: /enhancements/keps/sig-node/20181106-in-place-update-of-pod-resources.md

dashpole commented on pkg/kubelet/eviction/helpers.go:
```diff
@@ -562,12 +570,21 @@ func memory(stats statsFunc) cmpFunc {
 		p1Memory := memoryUsage(p1Stats.Memory)
-		p1Request := v1resource.GetResourceRequestQuantity(p1, v1.ResourceMemory)
+		if utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling) {
+			p1Request = v1resource.GetResourceAllocationQuantity(p1, v1.ResourceMemory)
```
I'd suggest checking the feature gate inside of this function, and returning GetResourceRequestQuantity if it isn't set, rather than having  all of these if/else statements.

vinaykul replied:
That's what I did at first, but when I read the code, it seemed to be hiding what's actually being returned. In particular, switching inside PodRequestsAndLimits worked fine for kubelet but changed the values seen by kubectl describe. We needed PodResourceAllocations, and after adding it, I felt that being explicit made it easier to read and understand what's going on. 

I'm worried that if we hide it inside, then down the road someone might use the function thinking they are getting Requests, while they are actually getting something else - and regress when we beta/GA this feature.

Once we GA and get rid of feature-gate, this will look clean.

Commit:
InPlace Pod Vertical Scaling feature - core implementation
1. Enable resource quota and limit ranger for pod resource changes.
2. In kubelet, do apiserver ResourcesAllocated patch under mutex.
KEP: /enhancements/keps/sig-node/20181106-in-place-update-of-pod-resources.md

Commit:
InPlace Pod Vertical Scaling feature - test fixes
1. Fix issues found by verify, typecheck, dependencies tests.
2. Fix error message during validation when InPlacePodVerticalScaling feature-gate is off.
3. Backout feature-gate change in kubeadm InitFeatureGates list.
KEP: /enhancements/keps/sig-node/20181106-in-place-update-of-pod-resources.md

Commit:
InPlace Pod Vertical Scaling feature - review fixes
1. Use strategicpatch to generate patch data for ResourcesAllocated when handling pod resize.
2. Use ContainersToStart instead of ContainersToRestart in SyncPod.
3. Simplify use of ContainersToUpdate in SyncPod.
KEP: /enhancements/keps/sig-node/20181106-in-place-update-of-pod-resources.md

Commit:
InPlace Pod Vertical Scaling feature - review fixes, initial e2e tests
1. Add framework code for e2e testing, and example e2e tests.
2. Do container resource update based on difference between requests as well as limits.
KEP: /enhancements/keps/sig-node/20181106-in-place-update-of-pod-resources.md

vinaykul commented:
This is stale. Working on implementation of updated design. Will be sending PRs to k/k in a couple of weeks once well tested.
